compiler/rustc_feature/src/builtin_attrs.rs

```diff
@@ -88,11 +89,70 @@ impl AttributeGate {
 /// Only top-level shape (`#[attr]` vs `#[attr(...)]` vs `#[attr = ...]`) is considered now.
 #[derive(Clone, Copy, Default)]
 pub struct AttributeTemplate {
+    /// If `true`, the attribute is allowed to be a bare word like `#[test]`.
     pub word: bool,
+    /// If `Some`, the attribute is allowed to take a list of items like `#[derive(..)]`.
     pub list: Option<&'static str>,
+    /// If `Some`, the attribute is allowed to be a name/value pair like `#[must_use = "reason"]`.
     pub name_value_str: Option<&'static str>,
 }
 
+/// How to handle multiple duplicate attributes on the same item.
+#[derive(Clone, Copy)]
+pub enum AttributeDuplicates {
+    /// Duplicates of this attribute are allowed.
+    ///
+    /// This should only be used with attributes where duplicates have semantic
+    /// meaning, or some kind of "additive" behavior. For example, `#[warn(..)]`
+    /// can be specified multiple times, and it combines all the entries. Or use
+    /// this if there is validation done elsewhere.
+    DuplicatesOk,
+    /// Duplicates after the first attribute will be an unused_attribute warning.
+    ///
+    /// This is usually used for "word" attributes, where they are used as a
+    /// boolean marker, like `#[used]`. It is not necessarily wrong that there
+    /// are duplicates, but the others should probably be removed.
+    WarnFollowing,
+    /// Same as `WarnFollowing`, but only issues warnings for word-style attributes.
+    ///
+    /// This is only for special cases, for example multiple `#[macro_use]` can
+    /// be warned, but multiple `#[macro_use(...)]` should not because the list
+    /// form has different meaning from the word form.
+    WarnFollowingWordOnly,
+    /// Duplicates after the first attribute will be an error.
+    ///
+    /// This should be used where duplicates would be ignored, but carry extra
+    /// meaning that could cause confusion. For example, `#[stable(since="1.0")]
+    /// #[stable(since="2.0")]`, which version should be used for `stable`?
+    ErrorFollowing,
+    /// Duplicates preceding the last instance of the attribute will be an error.
+    ///
+    /// This is the same as `ErrorFollowing`, except the last attribute is the
+    /// one that is "used". This is typically used in cases like codegen
+    /// attributes which usually only honor the last attribute.
+    ErrorPreceding,
```

Comment:
We shouldn't make this distinction, IMO, and should merge `ErrorFollowing` into this variant, and always use the last specified value. (And error on the previously specified values.)

Reply:
Yea, I was a little uneasy with having different following/preceding behaviors.

I can easily remove `ErrorFollowing`, but I feel like that emphasizes the inconsistencies:

* Normally when there are duplicate definitions (like items), the compiler generates an error about the following entries. Why would attributes have the opposite behavior?
* It seems inconsistent to have `WarnFollowing` and `ErrorPreceding`. Why have one behavior for warning and the opposite for errors?
    * Comparing to `unused_variables`, if you have `let x = 1; let x = 2; ...`, it will warn about the first `x`.

So, perhaps it would be better to remove `ErrorPreceding` instead? At least, I feel like that would be more consistent with the rest of the language. That would also match the behavior of things like the [proc-macro check](https://github.com/rust-lang/rust/blob/f31622a50b06e4e5775cf71d1dfe3e59ffc21aa5/compiler/rustc_builtin_macros/src/proc_macro_harness.rs#L194-L225) which generates errors for following duplicates.

Reply:
I mean if
```rust
#[key = "value1"]
#[key = "value2"]
NODE
```
was intentionally allowed, then `value2` would probably override `value1`, similarly to how later `let` or `macro_rules` "override" (shadow) earlier `let` or `macro_rules`, or how later command line options override earlier command line options.
So `value1` would be unused.

But this requires changing other parts of the compiler to harmonize the errors and the actual behavior after the error is reported, so it's probably not a material for this PR.